native/dlaexc.go

```diff
@@ -0,0 +1,267 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack"
+)
+
+// Dlaexc swaps two adjacent diagonal blocks of order 1 or 2 in an n×n upper
+// quasi-triangular matrix T by an orthogonal similarity transformation.
+//
+// T must be in Schur canonical form, that is, block upper triangular with 1×1
+// and 2×2 diagonal blocks; each 2×2 diagonal block has its diagonal elements
+// equal and its off-diagonal elements of opposite sign. On return, T will
+// contain the updated matrix again in Schur canonical form.
+//
+// If wantq is true, the transformation is accumulated in the n×n matrix Q,
+// otherwise Q is not referenced.
+//
+// j1 is the index of the first row of the first block. n1 and n2 are the order
+// of the first and second block, respectively.
+//
+// work much have length at least n, otherwise Dlaexc will panic.
+//
+// If ok is false, the transformed matrix T would be too far from Schur form.
+// The blocks are not swapped, and T and Q are not modified.
+func (impl Implementation) Dlaexc(wantq bool, n int, t []float64, ldt int, q []float64, ldq int, j1, n1, n2 int, work []float64) (ok bool) {
+	checkMatrix(n, n, t, ldt)
+	if wantq {
+		checkMatrix(n, n, q, ldq)
+	}
+	if j1 < 0 || n <= j1 {
+		panic("lapack: index j1 out of range")
+	}
+	if len(work) < n {
+		panic(badWork)
+	}
+	if n1 < 0 || 2 < n1 {
+		panic("lapack: invalid value of n1")
+	}
+	if n2 < 0 || 2 < n2 {
+		panic("lapack: invalid value of n2")
+	}
+
+	if n == 0 || n1 == 0 || n2 == 0 {
+		return true
+	}
+	if j1+n1 >= n {
```

Comment:
It's strange that Lapack checks whether the first block fits into the matrix but not the second. And moreover return true in that case?